src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -679,35 +680,155 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
+                    {
+                        // It should be harmless to have the recipient call MoveNext again.
+                        // but we guard against strange behaviour from shortcuts in the enumerator
+                        // It should also be harmless to dispose the enumerator here if we've
+                        // exhausted it, but we guard against strange behaviour in the caller.
+                        if (!sourceEnumerator.MoveNext()) return new DeadEnumeratorWrapper(sourceEnumerator);
+                    }
+                    return sourceEnumerator;
+                }
+                catch(Exception ex)
                 {
-                    while (e.MoveNext()) yield return e.Current;
+                    return new ErrorEnumeratorWrapper(sourceEnumerator, ExceptionDispatchInfo.Capture(ex));
                 }
             }
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
         }
 
         public static IEnumerable<TSource> SkipWhile<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
             if (source == null) throw Error.ArgumentNull("source");
             if (predicate == null) throw Error.ArgumentNull("predicate");
-            return SkipWhileIterator<TSource>(source, predicate);
+            return new SkipWhileIterator<TSource>(source, predicate);
         }
 
-        private static IEnumerable<TSource> SkipWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate)
+        private class SkipWhileIterator<TSource> : IEnumerable<TSource>
         {
-            bool yielding = false;
-            foreach (TSource element in source)
+            private class Enumerator : IEnumerator<TSource>
+            {
+                private readonly IEnumerator<TSource> _source;
+                private readonly Func<TSource, bool> _predicate;
+                private bool _started;
+                public Enumerator(IEnumerable<TSource> source, Func<TSource, bool> predicate)
+                {
+                    _source = source.GetEnumerator();
+                    _predicate = predicate;
+                }
+                public TSource Current
+                {
+                    get { return _source.Current; }
+                }
+                object IEnumerator.Current
+                {
+                    get { return Current; }
+                }
+                public bool MoveNext()
+                {
+                    if (_started)
+                        return _source.MoveNext();
+                    _started = true;
+                    while (_source.MoveNext())
+                        if (!_predicate(_source.Current))
+                            return true;
+                    return false;
+                }
+                public void Reset()
+                {
+                    _source.Reset();
+                    _started = false;
```

Comment:
I do not think former version supported Reset. It is better to not introduce it because in will or will not work depending on underlying source which would be inconsistent and hard to program against. More so, once it is supported it will be hard to take back if future optimizations cannot support Reset. 
Generally Linq iterators are not supporting reset. So it is better to not expand the feature surface here.